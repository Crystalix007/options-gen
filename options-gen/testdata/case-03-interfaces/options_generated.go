// Code generated by options-gen. DO NOT EDIT.
package testcase

import (
	"fmt"
	"io"

	"golang.org/x/sync/errgroup"
)

type optOptionsMeta struct {
	setter    func(o *Options)
	validator func(o *Options) error
}

func NewOptions(
	Any any,
	Stringer fmt.Stringer,
	RWCloser io.ReadWriteCloser,
	Local localInterface,

	options ...optOptionsMeta,
) Options {
	o := Options{}
	o.Any = Any
	o.Stringer = Stringer
	o.RWCloser = RWCloser
	o.Local = Local

	for i := range options {
		options[i].setter(&o)
	}

	return o
}

func WithOptAny(opt any) optOptionsMeta {
	return optOptionsMeta{
		setter:    func(o *Options) { o.OptAny = opt },
		validator: _Options_OptAnyValidator,
	}
}

func WithOptStringer(opt fmt.Stringer) optOptionsMeta {
	return optOptionsMeta{
		setter:    func(o *Options) { o.OptStringer = opt },
		validator: _Options_OptStringerValidator,
	}
}

func WithOptRWCloser(opt io.ReadWriteCloser) optOptionsMeta {
	return optOptionsMeta{
		setter:    func(o *Options) { o.OptRWCloser = opt },
		validator: _Options_OptRWCloserValidator,
	}
}

func WithOptLocal(opt localInterface) optOptionsMeta {
	return optOptionsMeta{
		setter:    func(o *Options) { o.OptLocal = opt },
		validator: _Options_OptLocalValidator,
	}
}

func (o *Options) Validate() error {
	g := new(errgroup.Group)

	g.Go(func() error {
		err := _Options_AnyValidator(o)
		if err != nil {
			return fmt.Errorf("invalid value for option WithAny: %w", err)
		}
		return nil
	})
	g.Go(func() error {
		err := _Options_StringerValidator(o)
		if err != nil {
			return fmt.Errorf("invalid value for option WithStringer: %w", err)
		}
		return nil
	})
	g.Go(func() error {
		err := _Options_RWCloserValidator(o)
		if err != nil {
			return fmt.Errorf("invalid value for option WithRWCloser: %w", err)
		}
		return nil
	})
	g.Go(func() error {
		err := _Options_LocalValidator(o)
		if err != nil {
			return fmt.Errorf("invalid value for option WithLocal: %w", err)
		}
		return nil
	})
	g.Go(func() error {
		err := _Options_OptAnyValidator(o)
		if err != nil {
			return fmt.Errorf("invalid value for option WithOptAny: %w", err)
		}
		return nil
	})
	g.Go(func() error {
		err := _Options_OptStringerValidator(o)
		if err != nil {
			return fmt.Errorf("invalid value for option WithOptStringer: %w", err)
		}
		return nil
	})
	g.Go(func() error {
		err := _Options_OptRWCloserValidator(o)
		if err != nil {
			return fmt.Errorf("invalid value for option WithOptRWCloser: %w", err)
		}
		return nil
	})
	g.Go(func() error {
		err := _Options_OptLocalValidator(o)
		if err != nil {
			return fmt.Errorf("invalid value for option WithOptLocal: %w", err)
		}
		return nil
	})
	return g.Wait()
}

func _Options_AnyValidator(o *Options) error {

	return nil
}

func _Options_StringerValidator(o *Options) error {

	return nil
}

func _Options_RWCloserValidator(o *Options) error {

	return nil
}

func _Options_LocalValidator(o *Options) error {

	return nil
}

func _Options_OptAnyValidator(o *Options) error {

	return nil
}

func _Options_OptStringerValidator(o *Options) error {

	return nil
}

func _Options_OptRWCloserValidator(o *Options) error {

	return nil
}

func _Options_OptLocalValidator(o *Options) error {

	return nil
}
