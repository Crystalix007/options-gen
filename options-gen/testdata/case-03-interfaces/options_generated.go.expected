// Code generated by options-gen. DO NOT EDIT.
package testcase

import (
	"fmt"
	"io"

	errors461e464ebed9 "github.com/kazhuravlev/options-gen/pkg/errors"
)

type optOptionsMeta struct {
	setter    func(o *Options)
	validator func(o *Options) error
}

func NewOptions(
	Any any,
	Stringer fmt.Stringer,
	RWCloser io.ReadWriteCloser,
	Local localInterface,

	options ...optOptionsMeta,
) Options {
	o := Options{}
	o.Any = Any
	o.Stringer = Stringer
	o.RWCloser = RWCloser
	o.Local = Local

	for i := range options {
		options[i].setter(&o)
	}

	return o
}

func WithOptAny(opt any) optOptionsMeta {
	return optOptionsMeta{
		setter:    func(o *Options) { o.OptAny = opt },
		validator: _Options_OptAnyValidator,
	}
}

func WithOptStringer(opt fmt.Stringer) optOptionsMeta {
	return optOptionsMeta{
		setter:    func(o *Options) { o.OptStringer = opt },
		validator: _Options_OptStringerValidator,
	}
}

func WithOptRWCloser(opt io.ReadWriteCloser) optOptionsMeta {
	return optOptionsMeta{
		setter:    func(o *Options) { o.OptRWCloser = opt },
		validator: _Options_OptRWCloserValidator,
	}
}

func WithOptLocal(opt localInterface) optOptionsMeta {
	return optOptionsMeta{
		setter:    func(o *Options) { o.OptLocal = opt },
		validator: _Options_OptLocalValidator,
	}
}

func (o *Options) Validate() error {
	errs := new(errors461e464ebed9.ValidationErrors)

	errs.Add(errors461e464ebed9.NewValidationError("Any", _Options_AnyValidator(o)))
	errs.Add(errors461e464ebed9.NewValidationError("Stringer", _Options_StringerValidator(o)))
	errs.Add(errors461e464ebed9.NewValidationError("RWCloser", _Options_RWCloserValidator(o)))
	errs.Add(errors461e464ebed9.NewValidationError("Local", _Options_LocalValidator(o)))
	errs.Add(errors461e464ebed9.NewValidationError("OptAny", _Options_OptAnyValidator(o)))
	errs.Add(errors461e464ebed9.NewValidationError("OptStringer", _Options_OptStringerValidator(o)))
	errs.Add(errors461e464ebed9.NewValidationError("OptRWCloser", _Options_OptRWCloserValidator(o)))
	errs.Add(errors461e464ebed9.NewValidationError("OptLocal", _Options_OptLocalValidator(o)))
	return errs.AsError()
}

func _Options_AnyValidator(o *Options) error {

	return nil
}

func _Options_StringerValidator(o *Options) error {

	return nil
}

func _Options_RWCloserValidator(o *Options) error {

	return nil
}

func _Options_LocalValidator(o *Options) error {

	return nil
}

func _Options_OptAnyValidator(o *Options) error {

	return nil
}

func _Options_OptStringerValidator(o *Options) error {

	return nil
}

func _Options_OptRWCloserValidator(o *Options) error {

	return nil
}

func _Options_OptLocalValidator(o *Options) error {

	return nil
}
